plugin/cache.go

```diff
@@ -227,14 +238,13 @@ func cachedOp(ctx context.Context, opName string, entry Entry, ttl time.Duration
 	}
 
 	if entry.id() == "" {
-		journal.Record(ctx, "Cached op %v on %v had no cache ID", opName, entry.name())
-		// This can happen if the entry's just been created, and it calls one of the Cached*
-		// methods to query its Metadata so that it can set its initial attributes. In that case,
-		// just invoke op. Ideally we should panic here, but panicking makes it significantly more
-		// annoying for plugin authors to write their code. Since the plugin package manages the
-		// entry's ID, and the entry's ID is set in CachedList, an empty ID is such a rare edge case
-		// that it's better to not annoy people.
-		return op()
+		// Try to set the ID based on parent ID
+		if obj := ctx.Value(parentID); obj != nil {
```

Comment:
Could we centralize this? Maybe change `e.setID` => `e.setIDFromParent`.

Reply:
The function would be a little awkward because we want to re-use a CName we've already calculated.